File: frontend/src/types/budget-alert.ts
export interface BudgetAlert {
  id: number;
  budget_id: number;
  category: string | null;
  alert_type: 'warning' | 'over_budget' | 'threshold_50' | 'threshold_80' | 'threshold_100' | 'bill_due';
  threshold_percentage: number;
  current_spending: number;
  budget_amount: number;
  amount_remaining: number | null;
  is_read: boolean;
  is_dismissed: boolean;
  created_at: string;
  budget_name?: string;
  bill_name?: string;
  amount?: number;
  message?: string;
  title?: string;
}

export interface BudgetThresholdStatus {
  budget_id: number;
  month_key: string;
  total_budget: number;
  total_spent: number;
  percentage_used: number;
  is_over_budget: boolean;
  thresholds: {
    [key: string]: {
      threshold_amount: number;
      is_exceeded: boolean;
      exceeded_at: string | null;
    };
  };
  category_status: CategoryThreshold[];
  next_threshold?: number;
}

export interface CategoryThreshold {
  category: string;
  budget_amount: number;
  spent: number;
  percentage: number;
  status: 'normal' | 'threshold_50' | 'threshold_80' | 'threshold_100' | 'over_budget';
}

export interface BudgetAlertListResponse {
  success: boolean;
  data: {
    alerts: BudgetAlert[];
    total_count: number;
    unread_count: number;
  };
  alerts: BudgetAlert[];
  total_count: number;
  unread_count: number;
}

File: backend/app/schemas/budget_alert.py
"""Budget alert schemas for API validation."""

from datetime import datetime
from decimal import Decimal
from pydantic import BaseModel, Field


class BudgetAlertCreate(BaseModel):
    """Schema for creating a budget alert."""

    user_id: int
    budget_id: int
    category: str | None = None
    alert_type: str = Field(
        ..., pattern="^(warning|over_budget|threshold_50|threshold_80|threshold_100)$"
    )
    threshold_percentage: Decimal = Field(..., ge=0, le=200)
    current_spending: int = Field(..., ge=0)
    budget_amount: int = Field(..., ge=0)
    amount_remaining: int | None = None


class BudgetAlertResponse(BaseModel):
    """Budget alert response."""

    id: int
    budget_id: int
    category: str | None = None
    alert_type: str
    threshold_percentage: Decimal
    current_spending: int
    budget_amount: int
    amount_remaining: int | None = None
    is_read: bool
    is_dismissed: bool
    created_at: datetime

    class Config:
        from_attributes = True


class BudgetAlertListResponse(BaseModel):
    """Response for list of budget alerts."""

    success: bool = True
    data: dict = Field(default_factory=lambda: {"alerts": [], "total_count": 0, "unread_count": 0})
    alerts: list[BudgetAlertResponse]
    total_count: int
    unread_count: int


class ThresholdStatus(BaseModel):
    """Status of a single threshold."""

    threshold_amount: int
    is_exceeded: bool
    exceeded_at: datetime | None = None


class CategoryThresholdStatus(BaseModel):
    """Threshold status for a category."""

    category: str
    budget_amount: int
    spent: int
    percentage: float
    status: str  # "normal", "threshold_50", "threshold_80", "threshold_100", "over_budget"


class BudgetThresholdStatusResponse(BaseModel):
    """Response for budget threshold status."""

    budget_id: int
    month_key: str
    total_budget: int
    total_spent: int
    percentage_used: float
    is_over_budget: bool
    thresholds: dict = Field(default_factory=dict)
    category_status: list[CategoryThresholdStatus]

    class Config:
        from_attributes = True


class MarkAlertReadRequest(BaseModel):
    """Request to mark alert as read."""

    pass


class MarkAllAlertsReadResponse(BaseModel):
    """Response for marking all alerts as read."""

    success: bool = True
    updated_count: int
